packages/connect/src/utils/debug.ts

```diff
@@ -14,63 +14,116 @@ const colors: Record<string, string> = {
     '@trezor/transport': 'color: #bada55; background: #000;',
 };
 
-type LogMessage = {
+export type LogMessage = {
     level: string;
     prefix: string;
     message: any[];
     timestamp: number;
+    css: string;
+};
+
+export type LogWriter = {
+    add: (message: LogMessage) => void;
 };
 
 const MAX_ENTRIES = 100;
 
+const stringify = (obj: Record<string, any>) => {
+    let cache: string[] = [];
+    const str = JSON.stringify(obj, (_key, value) => {
+        if (typeof value === 'object' && value !== null) {
+            if (cache.indexOf(value) !== -1) {
+                // Circular reference found, discard key
+                return;
+            }
+            // Store value in our collection
+            cache.push(value);
+        }
+        return value;
+    });
+    cache = [];
+    return str;
+};
+
 class Log {
     prefix: string;
     enabled: boolean;
     css: string;
     messages: LogMessage[];
+    logWriter: any;
 
-    constructor(prefix: string, enabled: boolean) {
+    constructor(prefix: string, enabled: boolean, logWriter?: LogWriter) {
         this.prefix = prefix;
         this.enabled = enabled;
         this.messages = [];
         this.css = typeof window !== 'undefined' && colors[prefix] ? colors[prefix] : '';
+        if (logWriter) {
+            this.logWriter = logWriter;
+        }
     }
 
-    addMessage(level: string, prefix: string, ...args: any[]) {
-        this.messages.push({
+    addMessage(level: string, prefix: string, css: string, ...args: any[]) {
+        const message = {
             level,
             prefix,
+            css,
             message: args,
             timestamp: Date.now(),
-        });
+        };
+
+        this.messages.push(message);
+
+        if (this.logWriter) {
+            const { level, prefix, timestamp, css, ...rest } = message;
+
+            // todo: this method calls post postMessage which serializes object passed in ...args.
+            // if there is cyclic dependency, it simply dies.
+            // this is probably not the right place to call stringify.
```

Comment:
imho validation of data should be in responsibility of logWriter itself, what do you think?